pkg/controllers/user/alert/watcher/cluster_scan.go

```diff
@@ -0,0 +1,161 @@
+	clusterAlertRuleLister v3.ClusterAlertRuleLister
+	alertManager           *manager.AlertManager
+}
+
+func StartClusterScanWatcher(ctx context.Context, cluster *config.UserContext, alertManager *manager.AlertManager) {
+
+	clusterLister := cluster.Management.Management.Clusters("").Controller().Lister()
+	clusterScans := cluster.Management.Management.ClusterScans(cluster.ClusterName)
+	clusterScanLister := clusterScans.Controller().Lister()
+	clusterAlertRuleLister := cluster.Management.Management.ClusterAlertRules(cluster.ClusterName).Controller().Lister()
+
+	clusterScanWatcher := &ClusterScanWatcher{
+		clusterName:            cluster.ClusterName,
+		clusterLister:          clusterLister,
+		clusterScanLister:      clusterScanLister,
+		clusterAlertRuleLister: clusterAlertRuleLister,
+		alertManager:           alertManager,
+	}
+	clusterScans.AddClusterScopedLifecycle(ctx, "cluster-scan-watcher", cluster.ClusterName, clusterScanWatcher)
+}
+
+func (csw *ClusterScanWatcher) Create(cs *v3.ClusterScan) (runtime.Object, error) {
+	return cs, nil
+}
+
+func (csw *ClusterScanWatcher) Remove(cs *v3.ClusterScan) (runtime.Object, error) {
+	return cs, nil
+}
+
+func (csw *ClusterScanWatcher) Updated(cs *v3.ClusterScan) (runtime.Object, error) {
+	if cs == nil {
+		return nil, nil
+	}
+	if cs.DeletionTimestamp != nil {
+		return cs, nil
+	}
+	// Start with Unknown, True if there is/are a matching alert rule(s), else False
+	if !(v3.ClusterScanConditionAlerted.IsUnknown(cs) &&
+		v3.ClusterScanConditionCompleted.IsTrue(cs)) {
+		return cs, nil
+	}
+	var err error
+	if csw.alertManager.IsDeploy == false {
+		logrus.Infof("ClusterScanWatcher: Sync: alert manager not deployed")
+		v3.ClusterScanConditionAlerted.False(cs)
+		v3.ClusterScanConditionAlerted.Message(cs, MsgAlertManagerNotDeployed)
+		return cs, nil
+	}
+	clusterAlertRules, err := csw.clusterAlertRuleLister.List("", labels.NewSelector())
+	if err != nil {
+		return cs, fmt.Errorf("ClusterScanWatcher: Sync: error listing cluster alert rules: %v", err)
+	}
+
+	match := false
+	var matchingAlertRules []*v3.ClusterAlertRule
+	for _, alertRule := range clusterAlertRules {
+		if alertRule.Status.AlertState == "inactive" || alertRule.Spec.ClusterScanRule == nil {
+			continue
+		}
+		if csw.isAlertRuleMatching(cs, alertRule) {
+			match = true
+			matchingAlertRules = append(matchingAlertRules, alertRule)
+		}
+	}
+	logrus.Debugf("ClusterScanWatcher: Sync: matchingAlertRules: %+v", matchingAlertRules)
+	if !match {
+		v3.ClusterScanConditionAlerted.False(cs)
+		v3.ClusterScanConditionAlerted.Message(cs, MsgNoMatchingAlertRule)
+		return cs, nil
+	}
+
+	alertSuccessful := true
+	for _, alertRule := range matchingAlertRules {
+		if e := csw.sendAlert(cs, alertRule); e != nil {
+			alertSuccessful = false
+			logrus.Errorf("ClusterScanWatcher: Sync: error sending alert: %v", e)
+			err = multierror.Append(err, e)
+		}
+	}
+	if !alertSuccessful {
+		return cs, err
```

Comment:
on error for any matchingAlertRule, we'd be sending alert again even for successful ones on next sync? is that expected?

Reply:
alert-manager kind of buffers the alerts for a short duration, this should be fine.